src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -351,6 +351,7 @@ trait TypeComparers {
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
     def isSub(ntp1: Type, ntp2: Type) = (ntp1.withoutAnnotations, ntp2.withoutAnnotations) match {
+      case (TypeRef(_, AnyClass, _), PolyType(_, _))                        => true                     // Any is kind-polymorphic
```

Comment:
Here's a test case (added to my branch) that exercises a more interesting code path (refining `M` to `Option`. The spurious error under 2.12.1 shows that the `TypeBounds` of two polytypes isn't automatically seen to be a polytype itself, hence the wrapping.

```
class Base[M[_]] { def mint: M[Int] = ??? }
final case class Sub() extends Base[Option]

class Test {
  def test[M[_]](b: Base[M]): Option[Int] = {
    bar : M[Any]
    b match {
      case Sub() =>
        bar : M[Any] // incorrect error: M does not take type parameters
        b.mint // GADT refinement: M = Option

        // Fix: change GADT refinement for HK type params to assign M the type:
        // [A]( >: Option[A] <: Option[A]),
        // rather than:
        // >: Option <: Option
        //
    }
  }
  def bar[N[_]]: N[Any] = ???
}
```

We have the same sort of bug in `lub` calculations, too.

```
scala> val tps = ListClass.typeConstructor :: OptionClass.typeConstructor :: Nil; val thelub = lub(tps)
tps: List[$r.intp.global.Type] = List(List, Option)
thelub: $r.intp.global.Type = Product with Serializable

scala> tps.forall(_.isHigherKinded)
res7: Boolean = true

scala> thelub.isHigherKinded
res8: Boolean = false
```